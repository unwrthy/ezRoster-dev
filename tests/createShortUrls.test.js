const createShortUrlsFactory = require('../src/createShortUrls')

jest.mock('request-promise-native', () => (options) => {
  const badRequestResponse = require('./fixtures/rebrandlyResponses/badRequest.json')
  const unauthorizedResponse = require('./fixtures/rebrandlyResponses/unauthorized.json')
  const invalidDomainResponse = require('./fixtures/rebrandlyResponses/invalidDomain.json')
  const invalidUrlResponse = require('./fixtures/rebrandlyResponses/invalidUrl.json')
  const serverErrorResponse = require('./fixtures/rebrandlyResponses/serverError.json')
  const serviceUnavailableResponse = require('./fixtures/rebrandlyResponses/serviceUnavailable.json')
  const successResponse = require('./fixtures/rebrandlyResponses/success.json')

  const body = JSON.parse(options.body)

  const error = new Error()
  if (options.headers.apikey === 'badRequest') {
    error.response = badRequestResponse
    return Promise.reject(error)
  } if (options.headers.apikey === 'invalidApiKey') {
    error.response = unauthorizedResponse
    return Promise.reject(error)
  } else if (options.headers.apikey === 'serverError') {
    error.response = serverErrorResponse
    return Promise.reject(error)
  } else if (options.headers.apikey === 'serviceUnavailable') {
    error.response = serviceUnavailableResponse
    return Promise.reject(error)
  } else if (body.destination === 'invalidUrl') {
    error.response = invalidUrlResponse
    return Promise.reject(error)
  } else if (body.domain && body.domain.fullName === 'invalidDomain') {
    error.response = invalidDomainResponse
    return Promise.reject(error)
  }

  return Promise.resolve(successResponse)
})

test('It should return a function', () => {
  expect(typeof createShortUrlsFactory('someValidApiKey')).toBe('function')
})

test('It should handle badRequest error', () => {
  const csu = createShortUrlsFactory('badRequest')
  expect(csu(['http://example.com'])).resolves.toMatchSnapshot()
})

test('It should handle invalid error because of an invalid apikey', () => {
  const csu = createShortUrlsFactory('invalidApiKey')
  expect(csu(['http://example.com'])).resolves.toMatchSnapshot()
})

test('It should handle error because of invalid url', () => {
  const csu = createShortUrlsFactory('someValidApiKey')
  expect(csu(['invalidUrl'])).resolves.toMatchSnapshot()
})

test('It should handle error because of invalid domain', () => {
  const csu = createShortUrlsFactory('someValidApiKey')
  expect(csu(['http://example.com'], 'invalidDomain')).resolves.toMatchSnapshot()
})

test('It should handle a generic server error', () => {
  const csu = createShortUrlsFactory('serverError')
  expect(csu(['http://example.com'])).resolves.toMatchSnapshot()
})

test('It should handle service unavailable error', () => {
  const csu = createShortUrlsFactory('serviceUnavailable')
  expect(csu(['http://example.com'])).resolves.toMatchSnapshot()
})

test('It should successfully create a shortUrl', () => {
  const csu = createShortUrlsFactory('someValidApiKey')
  expect(csu(['http://example.com'], 'somedomain.com', ['custom'])).resolves.toMatchSnapshot()
})
