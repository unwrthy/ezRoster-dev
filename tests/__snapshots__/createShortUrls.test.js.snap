// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should handle a generic server error 1`] = `
Array [
  [Error: Cannot create short URL for "http://example.com": Unexpected error connecting to Rebrandly APIs],
]
`;

exports[`It should handle badRequest error 1`] = `
Array [
  [Error: Cannot create short URL for "http://example.com": Bad Request],
]
`;

exports[`It should handle error because of invalid domain 1`] = `
Array [
  [Error: Cannot create short URL for "http://example.com": Not found: domain Not found],
]
`;

exports[`It should handle error because of invalid url 1`] = `
Array [
  [Error: Cannot create short URL for "invalidUrl": Invalid request: link Invalid format],
]
`;

exports[`It should handle invalid error because of an invalid apikey 1`] = `
Array [
  [Error: Cannot create short URL for "http://example.com": Unauthorized: Be sure you configured the integration to use a valid API key],
]
`;

exports[`It should handle service unavailable error 1`] = `
Array [
  [Error: Cannot create short URL for "http://example.com": Short URL service currently under maintenance. Retry later],
]
`;

exports[`It should successfully create a shortUrl 1`] = `
Array [
  Object {
    "clicks": 0,
    "createdAt": "2017-05-20T15:16:07.000Z",
    "creator": Object {
      "avatarUrl": "https://s.gravatar.com/avatar/e18d2c2510d828bc95bde6817e6e9ce2?size=80&d=retro&rating=g",
      "fullName": "Luciano",
      "id": "25bbd054d27e408f8ab9e0e1dd8de112",
    },
    "destination": "https://facebook.github.io/jest/docs/expect.html",
    "domain": Object {
      "fullName": "rebrand.ly",
      "id": "8f104cc5b6ee4a4ba7897b06ac2ddcfb",
      "ref": "/domains/8f104cc5b6ee4a4ba7897b06ac2ddcfb",
    },
    "domainId": "8f104cc5b6ee4a4ba7897b06ac2ddcfb",
    "domainName": "rebrand.ly",
    "favourite": false,
    "forwardParameters": true,
    "https": true,
    "id": "1611678",
    "isPublic": false,
    "linkId": 1611678,
    "scripts": Array [],
    "shortUrl": "rebrand.ly/dxws",
    "slashtag": "dxws",
    "status": "active",
    "tags": Array [],
    "title": "Link dxws",
    "updatedAt": "2017-05-20T15:16:07.000Z",
  },
]
`;
